Source/DafnyCore/AST/Tokens.cs

```diff
@@ -145,6 +169,10 @@ public class RangeToken : TokenWrapper {
   public RangeToken(IToken startTok, IToken endTok) : base(startTok) {
     this.endTok = endTok;
   }
+
+  public override IToken WithVal(string newVal) {
+    return this;
```

Comment:
> How is this a valid implementation of WithVal?

No one is supposed to change the "val" of a RangeToken. Would you prefer me to throw an exception? I hate exceptions if they are not necessary, because otherwise, things are covered with tests anyway.

This comes from the bad design that we are using tokens both for tokens, for positions AND for ranges.

Reply:
> Yes this clas is used by the Parser (see Dafny.atg)

Seems like only the `Fill` method is used. Could you delete `DetectNewline` ?

Also, could you add documentation to `PreprocessorHelper`?

Reply:
> This comes from the bad design that we are using tokens both for tokens, for positions AND for ranges.

Happy to hear you say that ^^ I am boggled by our overloaded use of `Token`

Reply:
> Seems like only the Fill method is used. Could you delete DetectNewline ?

Knowing that DetectNewline is defined and used in this PR?

> Also, could you add documentation to PreprocessorHelper?
Done

> Happy to hear you say that ^^ I am boggled by our overloaded use of Token

Not to mention... include tokens?! refinement tokens?! We should have kept tokens minimal and created a Provenance class that contains a regular token, a possible chain of module names, a possible chain of included files, etc. But the reason why we have everything in tokens is that it's the only formal information we passed to Boogie so far. This could change in the future because we have proof obligation descriptions which can be used to be return better error messages. They could be built using provenance.